### PR TITLE
Fetch validators from consensus contract

### PR DESCRIPTION
**Pull request by callistonianembrace**
## Depends
- [x] #157 

## Problem
We manually generate validators.json with sh scripts on the client, where as the deployed `consensus.mligo` should be the source of truth.

## Solution
This PR uses the Taquito script from #157 to fetch the validators set

### CODE REVIEW

**Review by anchpop — APPROVED** [review]

File: bin/http_server.re [hunk 1]
```diff
@@ -177,10 +177,36 @@ let handle_ticket_balance =
...
+             | Some(address) => address
+             | None =>
+               failwith(
+                 Printf.sprintf(
+                   "Could not parse %s received from the contract into an address",
+                   validator,
+                 ),
+               )
+             },
+             Printf.sprintf("https://localhost:444%d", i) |> Uri.of_string,
```
Comment (callistonianembrace) [inline bin/http_server.re:204]: Opened a follow up issue to keep this PR self contained
https://github.com/marigold-dev/sidechain/issues/160

File: tezos_interop/tezos_interop.rei [hunk 2]
```diff
@@ -154,3 +154,9 @@ module Consensus: {
 };
 
 module Discovery: {let sign: (Secret.t, ~nonce: int64, Uri.t) => Signature.t;};
+
+module Fetch_validators: {
+  let run:
+    (~rpc_node: string, ~contract_address: string) =>
+    Lwt.t(result(list(string), string));
+};
```
Comment (EduardoRFS) [inline tezos_interop/tezos_interop.rei:162]: This should not exported, look how the other interops with taquito do. TLDR, it should be exported as part of the Consensus module and not use strings for types.

File: tezos_interop/tezos_interop.re [hunk 3]
```diff
@@ -539,10 +539,10 @@ module Fetch_validators: {
       rpc_node: Uri.to_string(rpc_node),
       confirmation,
       contract_address: Address.to_string(contract_address),
-    };
+      };
```
Comment (EduardoRFS) [inline tezos_interop/tezos_interop.re:542]: Not needed spaces

File: tezos_interop/tezos_interop.re [hunk 4]
```diff
@@ -539,10 +539,10 @@ module Fetch_validators: {
...
     // TODO: stop hard coding this
     let command = "node";
-    let.await file =
+      let.await file =
```
Comment (EduardoRFS) [inline tezos_interop/tezos_interop.re:545]: Not needed spaces

File: tezos_interop/tezos_interop.rei [hunk 5]
```diff
@@ -151,6 +153,8 @@ module Consensus: {
   };
   let listen_operations:
     (~context: Context.t, ~on_operation: operation => unit) => unit;
+  let fetch_validators:
+    (~context: Context.t) => Lwt.t(result(list(Secret.t), string));
```
Comment (EduardoRFS) [inline tezos_interop/tezos_interop.rei:157]:
```suggestion
    (~context: Context.t) => Lwt.t(result(list(Key.t), string));
```

There is no way the storage can return `Secret.t`, as they're never stored on Tezos to begin with, probably this should be Key.t

File: tezos_interop/tezos_interop.re [hunk 6]
```diff
@@ -748,6 +790,52 @@ module Consensus = {
...
+        }
+      | Ok(_) => Error("Failed to parse storage micheline expression")
+      | Error(msg) => Error(msg);
+    let.await micheline_storage =
+      Fetch_validators.run(
+        ~confirmation=required_confirmations,
+        ~rpc_node,
+        ~contract_address=consensus_contract,
+      );
+    Lwt.return @@ micheline_to_validators(micheline_storage);
```
Comment (EduardoRFS) [inline tezos_interop/tezos_interop.re:837]:
```suggestion
    Lwt.return(micheline_to_validators(micheline_storage));
```

File: tezos_interop/tezos_interop.re [hunk 7]
```diff
@@ -748,6 +790,52 @@ module Consensus = {
...
+        ) => {
+          Ok(
+            List.map(
+              fun
+              | Micheline.String(_, sk) =>
+                switch (Secret.of_string(sk)) {
+                | Some(sk) => sk
+                | None => failwith("sk isn't a valid Secret")
+                }
+              | _ => failwith("Should have been String"),
```
Comment (EduardoRFS) [inline tezos_interop/tezos_interop.re:824]: This should not be a failwith, I think there is a `fold_left_ok` if you want you can do a make `map_ok`

**Review by EduardoRFS — CHANGES_REQUESTED** [review]

File: tezos_interop/tezos_interop.re (hunk 7 above)
Comment (callistonianembrace) [reply tezos_interop/tezos_interop.re]: Didn't get you - what's to be done with keys that dont parse?

Comment (EduardoRFS) [reply tezos_interop/tezos_interop.re]: return an Error()

Comment (callistonianembrace) [reply tezos_interop/tezos_interop.re]: We could still use the keys that parse - what about a filter_map?

Comment (EduardoRFS) [reply tezos_interop/tezos_interop.re]: If a single key failed then everything needs to fail.

File: tezos_interop/tezos_interop.re [hunk 8]
```diff
@@ -806,6 +806,51 @@ module Consensus = {
...
+                _,
+                D_Pair,
+                [_, Prim(_, D_Pair, [_, Prim(_, D_Pair, keys, _)], _)],
+                _,
+              ),
+            ],
+            _,
+          ),
+        ) => {
+          List.fold_left(
```
Comment (EduardoRFS) [inline tezos_interop/tezos_interop.re:829]: You should use `fold_left_ok` here

**Review by EduardoRFS — CHANGES_REQUESTED** [review]

File: tezos_interop/fetch_storage.js [hunk 9]
```diff
@@ -60,5 +60,7 @@ const error = (error) =>
     } else {
       error({ message: "Not in current Branch" });
     }
+  } else {
+    error({ message: "Internal error: operationFromHead was undefined" });
```
Comment (EduardoRFS) [inline tezos_interop/fetch_storage.js:64]: That's a nice catch

File: tezos_interop/tezos_interop.re [hunk 10]
```diff
@@ -555,29 +555,26 @@ module Fetch_storage: {
...
+      let.await output =
+        Lwt_process.pmap(
+          (command, [|command, file|]),
+          Yojson.Safe.to_string(input_to_yojson(input)),
+        );
+      switch (Yojson.Safe.from_string(output) |> output_of_yojson) {
+      | Ok(storage) => await(Ok(storage))
+      | Error(error) => await(Error(error))
+      };
+    });
```
Comment (EduardoRFS) [inline tezos_interop/tezos_interop.re:577]: I don't think this change should be here, also I noticed a problem that this file is being created on every execution and not in a global level.

**Review by EduardoRFS — CHANGES_REQUESTED** [review]

File: tezos_interop/tezos_interop.re (hunk 10 above)
Comment (callistonianembrace) [reply tezos_interop/tezos_interop.re]: Addressed by rebasing in the next commit

File: tezos_interop/tezos_interop.rei [hunk 11]
```diff
@@ -102,6 +102,8 @@ module Context: {
     required_confirmations: int,
   };
 };
+
+type michelson = Micheline.node(int, Michelson_v1_primitives.prim);
```
Comment (EduardoRFS) [inline tezos_interop/tezos_interop.rei:106]:
```suggestion
```

File: tezos_interop/tezos_interop.re [hunk 12]
```diff
@@ -808,6 +809,42 @@ module Consensus = {
...
+            [Prim(_, D_Pair, [_, Seq(_, keys)], _), _, _],
+            _,
+          ),
+        ) => {
+          fold_left_ok(
+            (acc, k) =>
+              switch (k) {
+              | Micheline.String(_, k) =>
+                switch (Key.of_string(k)) {
+                | Some(k) => Ok([k, ...acc])
```
Comment (EduardoRFS) [inline tezos_interop/tezos_interop.re:829]: This reverts the order of validators, and it shouldn't, the order of validators is quite important for the consensus, as it's how we decide on the next validator.

**Review by EduardoRFS — CHANGES_REQUESTED** [review]

File: tezos_interop/tezos_interop.rei (hunk 11 above)
Comment (callistonianembrace) [reply tezos_interop/tezos_interop.rei:106]: What do you want here exactly?

Comment (callistonianembrace) [reply tezos_interop/tezos_interop.rei]: Attempted addressing this in the incoming commit

File: tezos_interop/tezos_interop.re (hunk 12 above)
Comment (callistonianembrace) [reply tezos_interop/tezos_interop.re:833]: Added a List.rev

File: tezos_interop/tezos_interop.re [hunk 13]
```diff
@@ -507,8 +507,6 @@ module Run_contract = {
   };
 };
 
-type michelson =
-  Tezos_micheline.Micheline.node(int, Michelson_v1_primitives.prim);
```
Comment (EduardoRFS) [inline tezos_interop/tezos_interop.re:511]: Why did you remove this? Now the type is being duplicated in 3 places, please revert this change

**Review by EduardoRFS — CHANGES_REQUESTED** [review]

File: tezos_interop/tezos_interop.re (hunk 13 above)
Comment (callistonianembrace) [reply tezos_interop/tezos_interop.re:511]: https://github.com/marigold-dev/sidechain/pull/158#discussion_r705558502

**Review by EduardoRFS — APPROVED** [review]
anime girl happy with the PR getting merged
